KEP-4819: Container log rotation on Disk perssure

Zeel-Patel opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description:

<!-- link to the k/enhancements issue -->
- Issue link:

<!-- other comments or additional information -->
- Other comments:

Zeel-Patel commented:
/sig-node

kannon92 commented on keps/sig-node/4819-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,451 @@
+
+## Design Details
+
+Define 2 new flags `logRotateDiskCheckInterval`, `logRotateDiskPressureThreshold` in kubelet config.
```
I was thinking that when there is disk pressure in nodefs.available/imagefs.available we can add rotateLogs to our list of functions we run in case of disk pressure. 

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/helpers.go#L1195-#L1230

Zeel-Patel replied:
That's better. Thanks.
But won't that use kubelet eviction thresholds in that case?

harshanarayana replied:
I suppose this log cleanup is another form of eviction to make space like we do with Images and containers. 

So, Adding that to the `nodefs.available` / `imagefs.available` would logically make a great addition

ffromani replied:
big +1, this was my initial thinking as well. Detecting  disk pressure should be a trigger for log rotation in addition to the timer-based approach.

Zeel-Patel replied:
Sure.

Considering this 
https://github.com/kubernetes/enhancements/pull/5022#discussion_r1906606844

If we rely on same kubelet eviction thresholds, it will cause log rotation to fail. We have experienced it as well, that log copy creation fails.

Zeel-Patel replied:
@ffromani @harshanarayana @kannon92 

Listing down few options, lmk your thoughts
- We use same disk eviction thresholds. Every time there is disk pressure, check size of .gz log files of all containers. If the combined file size (for a particular container) exceeds (containerLogMaxSize*containerLogMaxFiles), simply delete older .gz files till remaining size is within this limit. No log rotate. So if we have  containerLogMaxSize = 20Mib and containerLogMaxFiles = 6, we will ensure that combined log file size remains within 120Mib. It could also happen that a single .gz file endsup taking all 120Mib after cleanup
- Blindly do log rotation as proposed in KEP (But using disk pressure thresholds as suggested by you all)
- Make the change suggested in option 1 in ContainerLogManager periodic log rotation itself.

kannon92 replied:
my suggestion is when disk pressure is hit (by our eviction settings) we would trigger log rotation of all containers.

The goal would be to preemptly try and clean up disk space without evicting.

Zeel-Patel replied:
Looking at this goal, this seems good to me

We use same disk eviction thresholds. Every time there is disk pressure, check size of .gz log files of all containers. If the combined file size (for a particular container) exceeds (containerLogMaxSize*containerLogMaxFiles), simply delete older .gz files till remaining size is within this limit. No log rotate. So if we have containerLogMaxSize = 20Mib and containerLogMaxFiles = 6, we will ensure that combined log file size remains within 120Mib. It could also happen that a single .gz file endsup taking all 120Mib after cleanup

kannon92 commented:
/ok-to-test

harshanarayana commented on keps/sig-node/4819-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,451 @@
+
+Define 2 new flags `logRotateDiskCheckInterval`, `logRotateDiskPressureThreshold` in kubelet config.
+
+- `logRotateDiskCheckInterval` is the time interval within which the ContainerLogManager will check Disk usage on the kubelet host.
```
Here is my two cent. 

I was the one who added a similar config to the Log rotate workflow where it can take a configurable number workers with a duration and rotate logs in async mode to help reduce the leak of log file size when the container log generation rate it way too high. 

Which definitely helps, but never really prevents the issue. The safest way to deal with this would be to truncate and fix the size when the logs are being written instead of monitor, rotate and cleanup. That will be the most fool proof way to do this. 

That being said, In absence of that workflow, I like your suggestion.

> it will rotate logs of all the containers of the kubelet.

However, I am not sure if all service/containers should be taxed because of one service's misbehaviour.  The reason being how the `kubectl log` would look like after one such global rotation. 

As far as I can recall, the `.gz` extensions are ignored when the `kubectl logs` are being used. So, forcing a global rotation can lead to logs going missing from the `kubectl log` out of nowhere. 

We should still cleanup only those logs that have exceeded the configured threshold in this loop so that we let the other service behave the way they do.

Zeel-Patel replied:
Yes, that's what we are proposing.
Currently, log files are anyways exceeding the set thresholds which causes high disk usage.
Instead of evicting pods, we can once try rotating logs of such containers (containers for which logs size is exceeding the containerLogMaxSize threshold)

harshanarayana commented on keps/sig-node/4819-log-rotate-on-disk-pressure/README.md:
````diff
@@ -0,0 +1,451 @@
+It often happens that the containers generating heavy log data have compressed log file with size exceeding the containerLogMaxSize limit set in kubelet config.
+
+For example, kubelet has
+```
+containerLogMaxSize = 200M
+containerLogMaxFiles = 6
+```
````
FYI, there is also two other fields to configure the worker + the monitoring duration to perform the cleanup. 

`containerLogMaxWorkers` and `containerLogMonitorInterval` respectively.

harshanarayana commented on keps/sig-node/4819-log-rotate-on-disk-pressure/README.md:
````diff
@@ -0,0 +1,451 @@
+-rw-r--r-- 1 root root 183336376 Jan  1 11:29 0.log.20250101-112616.gz
+-rw-r----- 1 root root 205360966 Jan  1 11:29 0.log.20250101-112911
+```
+
+
+If the pod had been generating logs in Gigabytes with minimal delay, it can cause disk pressure on kubelet host and that can affect other pods running in the same kubelets.
+
````
This can also prevent further container log rotation to fail and make the problem even worse. 

case and point, an issue we ran into very recently. 


> Dec 11 21:35:31 tardis-node-192-168-4-51 kubelet[8204]: E1211 21:35:31.794811    8204 container_log_manager.go:296] "Failed to rotate log for container" err="failed to compress log \"/var/log/pods/tardis-pod-2-5bb1ae31-tm-78475757b8-2mbl9_f53b5f87-5b48-4748-a919-9eb702bb5f0c/taskmanager/2.log.20241211-202335\": failed to create temporary log \"/var/log/pods/tardis-pod-2-5bb1ae31-tm-78475757b8-2mbl9_f53b5f87-5b48-4748-a919-9eb702bb5f0c/taskmanager/2.log.20241211-202335.tmp\": open /var/log/pods/tardis-pod-2-5bb1ae31-tm-78475757b8-2mbl9_f53b5f87-5b48-4748-a919-9eb702bb5f0c/taskmanager/2.log.20241211-202335.tmp: disk quota exceeded" path="/var/log/pods/tardis-pod-2-5bb1ae31-tm-78475757b8-2mbl9_f53b5f87-5b48-4748-a919-9eb702bb5f0c/taskmanager/2.log" containerID="2c7de6fdcbda4498543ab3cc68d59dcee487e08ad7d37751f9ea5c366975f784"

Zeel-Patel replied:
I see, but the periodic log rotations is anyways failing.
If we wisely chose logRotateDiskPressureThreshold, it would be helpful.
wdyt?

harshanarayana replied:
It should definitely help. But the way to chose the right size and frequency to monitor is trial and error and I can't think of an effective way to determine the number. But that is besides the scope of this KEP anyway.

harshanarayana commented on keps/sig-node/4819-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,451 @@
+
+### Goals
+
+- Rotate and Clean all container logs on kubelet Disk pressure
```
Would there be a configuration required to help identify what logs are to be selected for such cleanup ?

Zeel-Patel replied:
Containers for which log size is exceeding containerLogMaxSize will be selected for cleanup.

harshanarayana reviewed (commented):
Leaving behind a few comments/ Feel free to ignore if you consider them irrelevant/invalid.

ffromani commented on keps/sig-node/4819-log-rotate-on-disk-pressure/kep.yaml:
```diff
@@ -0,0 +1,15 @@
+editor: "@Zeel-Patel"
+creation-date: 2025-01-08
+last-updated: 2025-01-08
+reviewers:
```
please add @kannon92 and @ffromani for starters

ffromani commented on keps/sig-node/4819-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,451 @@
+
+Just because a KEP is merged does not mean it is complete or approved. Any KEP
+marked as `provisional` is a working document and subject to change. You can
+denote sections that are under active debate as follows:
```
feel free to remove comments when no longer needed, to make review (and reading in general) easier

ffromani commented on keps/sig-node/4819-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,451 @@
+
+Rotate containers logs when there is disk pressure on kubelet host.
+
+## Motivation
```
we will need something here

ffromani commented on keps/sig-node/4819-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,451 @@
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
```
we will need some details here

ffromani replied:
brainstorming, let's evaluate some ideas:
- if the log rotation manager detects that the previous X rotations (1? 2? N?) made the rotated logs exceeds the configured max size, because containers are producing many logs continuously, should this cause kubelet to report disk pressure?
- do we want or need to review log retention also? if a rotated log is like 500 megs while it is supposed to be 100 megs (random example numbers), should _that_ cause the kubelet to report disk pressure? arguably, we are consuming more disk space than the user configured, and thus expects.

harshanarayana replied:
I am not sure if these should trigger a disk pressure report. That can have a side-effect. Right ?

However, this can definitely be Turned into a warning event from the node level to indicate that something is going wrong. Would that lead to an event storm for some cases? Depends on how frequently we generate the event.

harshanarayana replied:
> if the log rotation manager detects that the previous X rotations (1? 2? N?) made the rotated logs exceeds the configured max size

This problem becomes even more concerning with high value being set for `containerLogMaxFiles`.

ffromani commented on keps/sig-node/129447-log-rotate-on-disk-pressure/kep.yaml:
```diff
@@ -0,0 +1,15 @@
+editor: "@Zeel-Patel"
+creation-date: 2025-01-08
+last-updated: 2025-01-08
+reviewers:
```
please also add @harshanarayana

ffromani commented on keps/sig-node/129447-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,451 @@
+
+### Risks and Mitigations
+
+No identified risk.
```
let's think a bit more about this, especially if we start from beta. What can go wrong?

ffromani commented on keps/sig-node/129447-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,451 @@
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane? Yes (kubelet restart)
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? Yes
```
will it? why exactly? could you please elaborate?

harshanarayana replied:
1. Suppose this is a function of what your cloud provider lets you do in terms of tuning the kubelet configuration. 
2. For on prem deployments, this definitely is just a kubelet restart away. So, not much of a downtime/re-provisioning

ffromani replied:
The cloud provider use case is interesting indeed, but from project perspective this looks like the BM use case anyway. I think a good first step is just adding more details on the answer here.

kannon92 commented:
@Zeel-Patel Please create an issue in kubernetes/enhancement that tracks this KEP. That would be the title of the KEP.

The k/k issue is used as a discussion to create the KEP. Sorry to be pain about this.

harshanarayana commented on keps/sig-node/129447-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,451 @@
+
+- Rotate and Clean all container logs on kubelet Disk pressure
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
```
Some of the details from the KEP should go into this section.

harshanarayana commented on keps/sig-node/129447-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,451 @@
+No
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+ContainerLogManager of kubelet will use more CPU cycle then now.
```
```suggestion
CPU cycles usage of ContainerLogManager of kubelet will increase.
```

kannon92 reviewed (commented):
I worry that this won't work out that well when we come close to disk pressure. Let's say that a few pods are exceeding their limits and then we trigger a function to rotate. If we are close to host file system limits would log rotation still be expected to work?

ffromani commented:
/retitle KEP-5032: Container log rotation on Disk pressure

sftim commented on keps/sig-node/5032-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,305 @@
+
+## Motivation
+
+- A lot of out kubelet hosts experienced Disk pressure as a certain set of pods was generating very high logs. The rate was around 3-4Gib in 15 minutes. We had containerLogMaxSize set to 200Mib and containerLogMaxFiles set to 6. But the .gz files were of size around 500-600Gib. We observed that container log rotation was slow for us.
```
In this write up, who is "our"? Kubernetes project?

Zeel-Patel replied:
Updated

sftim commented on keps/sig-node/5032-log-rotate-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,305 @@
+- Rotate and Clean all container logs on kubelet Disk pressure that has exceeded the configured log retention quota
+
+## Proposal
+
+
```
There's no proposal. You should add one.

sftim commented on keps/sig-node/5032-log-rotate-on-disk-pressure/kep.yaml:
```diff
@@ -0,0 +1,17 @@
+title: Log rotate on Disk pressure
+kep-number: TBD
```
5032

Zeel-Patel commented:
/sig node

harshanarayana commented:
@kannon92 / @sftim 

He is a long rant/response to very valid point that both of you brought up. 

I think the source of the problem exists in the fact that kubelet is the one rotating after the longs are written onto the disk. So, no matter how hard we try, there will always be a possibility for leak. Since we are monitoring and reacting after the fact. The duration can be turned to as small a value as possible and tax kubelet, but that will also not be fool proof. The safest would be to truncate the log to size in the writer and not write the logs higher than configured size to the host. Drop everything else. But that also comes with a lot of problem. What if we end up having partial logs. That can break a lot of tooling and workflows. 

> provide a means for an external tool to trigger the kubelet to rotate its logs. That would move the policy decisions outside of the kubelet; for example, into a DaemonSet.

We tried this out internally, and had the same behavior more or less as that of doing in the kubelet. It did give us more configure knobs and options, but the underlying matter is, we are doing the cleanup after the logs are written first. 

So, we are equally failure prone is it not ?

> what I am trying to get is if we "fix" this bug of log rotations exceeding limits

This KEP and my old PR both did part of that work. But never got it fully accurate. The crux is in how we cleanup. 

Say you have a heavy logging pod, that is writing 10G per second (Silly number, but hey! Where is the fun otherwise). Say your configured limit was 100M per pod with 5 files. The one file written will easily exceed all configured limits. 

Now the question is, how do you cleanup these files. Do you cleanup the entire file? Or part of the file ? Both of whcih can be really flaky and very opinionated. Can cause loss of logs in both cases.

So, when my original change was done, I left the truncating part of the file out and cleaned up exceed files based on count alone. That fixed part of the problem But then the current issue that @Zeel-Patel  mentioned happen. Now comes the part of another cleanup that we have to where we sweep across and find all exceed limits and cleanup. But this cleanup is bound by the same problem that my original change would have had to deal with. 

Honestly, the safest I can think of for mitigating this would be to find a mechanism to truncate at the source ensuring we do not run into partial logs being written and then kubelet can just do the count based cleanup.

pohly commented:
> Honestly, the safest I can think of for mitigating this would be to find a mechanism to truncate at the source ensuring we do not run into partial logs being written

This is only safe if the process which writes the log data is also the one which rotates because it knows the message boundaries. But we removed log rotation from klog. There is now a proposal in https://github.com/kubernetes/kubernetes/pull/127667 to add it back. Perhaps something could be done with datagram stream sockets (preserve message boundaries) there.

sftim commented:
AIUI there is no safe way to have a process ship logs; it is _always_ possible to have a log volume that fills either local storage or a local buffer. There's no safe means because the rate of application log writes is formally not bounded, whereas buffers and local storage are always finite.

If we provide a way for container log writes to go to a socket or other file descriptor not backed by local storage, there is no need to rotate. That approach doesn't prevent log entries being dropped but it is proof against any local storage exhaustion.

harshanarayana commented:
@pohly I noticed that PR a couple of days ago but how does that help for service that do not really use that bin to run their process? For such, it is containerd that is doing the write. Correct ? https://github.com/containerd/containerd/blob/main/internal/cri/io/logger.go (is my understanding of this incorrect? )

For example, containerd exposes a knob `max_container_log_line_size`, would it make this better if the rotation were to be done there? Would that give us a slight bit more precision on how accurately we meet the log size per file configuration? 

> If we provide a way for container log writes to go to a socket or other file descriptor not backed by local storage, there is no need to rotate.

@sftim  Most definitely, But this might not always be possible for everyone to configure even it provided as an option. I still remember the days when docker had a whole slew of plugins one can configure for log management that could help avoid writing to the local storage. Or you could create one yourself.

pohly commented:
> I noticed that PR a couple of days ago but how does that help for service that do not really use that bin to run their process?

I doesn't. An app has to be configured explicitly to redirect output to its own files and then not write anything to stdout/stderr... which breaks `kubectl logs`. I still need to review that PR again in more detail, your comments on the design and usage would be very welcome.

sftim commented:
>  whole slew of plugins one can configure for log management

I'd definitely support a KEP to make container logging more extensible / pluggable. The story we have now is OK but not great.

Zeel-Patel commented:
> Honestly, the safest I can think of for mitigating this would be to find a mechanism to truncate at the source ensuring we do not run into partial logs being written and then kubelet can just do the count based cleanup.


- On disk pressure, analyse log paths of all containers. Logs of containers with existing combined log size exceeding `containerLogMaxSize`*`containerLogMaxFiles` should be deleted till the combined log size is within `containerLogMaxSize`*`containerLogMaxFiles`.

what do you think about this proposal?

Zeel-Patel commented:
@sftim @pohly @harshanarayana @kannon92 @ffromani 
Hi folks, does this KEP need any change to get approval or are we not planning to make any change?

harshanarayana commented:
> On disk pressure, analyse log paths of all containers. Logs of containers with existing combined log size exceeding containerLogMaxSize * containerLogMaxFiles should be deleted till the combined log size is within containerLogMaxSize * containerLogMaxFiles.

Yes. But that could also mean that we might end of removing all logs from some container. Right ?

Say a service is generating 10G per second data and the configured retention was 3 files with 100M each. So, if your log files are all above 300M(which is very likely to happens based on a very high log write rate), you will end up removing all 3 files, Which might be against the expected behavior. How do we protect against such cleanup ? Or should we even protect against such cleanup ?

harshanarayana commented:
> I'd definitely support a KEP to make container logging more extensible / pluggable. The story we have now is OK but not great.

@sftim  Would this be a kubelet side of configuration or would this be a containerd side of the config :? Would fit more into containerd side right ? Say a configurable log write destination that is not just file based/a more controlled way to write to a file based destination with some more guard rails ?

sftim commented:
For example, kubelet could send a memfd to the container runtime per container, which gets used as a ring buffer.

(handily, the kubelet can pass a read-only capability for the same memfd to a log helper daemon).

**There are lots of viable designs.**

Zeel-Patel commented:
> > On disk pressure, analyse log paths of all containers. Logs of containers with existing combined log size exceeding containerLogMaxSize * containerLogMaxFiles should be deleted till the combined log size is within containerLogMaxSize * containerLogMaxFiles.
> 
> @Zeel-Patel Yes. But that could also mean that we might end of removing all logs from some container. Right ?
> 
> Say a service is generating 10G per second data and the configured retention was 3 files with 100M each. So, if your log files are all above 300M(which is very likely to happens based on a very high log write rate), you will end up removing all 3 files, Which might be against the expected behavior. How do we protect against such cleanup ? Or should we even protect against such cleanup ?

Yes, I agree this solution is hacky and is not guaranteed to always work.

kannon92 commented:
Hey @Zeel-Patel!

Thank you for leading this effort. I think the story around log management in Kubelet needs some thought. We have had quite a few proposals around improving this.

Some contributors have asked that we disable log rotation entirely (ie https://github.com/kubernetes/kubernetes/issues/123279 from @rikatz). We also had the work from @harshanarayana to add parallel workers to help maintain log rotations. And now we have your request on adding log rotation on disk pressure. I know that there is also interest in moving logs to a separate partition from kubelet but not sure if there is tracking work for that.

We briefly talked about this in our sig-node planning call with @dchen1107 that our story around log management in Kubernetes is not that clear. I think this can be a topic for sig-node and we should discuss our stance on this.

Zeel-Patel commented:
> /hold
> 
> Let's meet and discuss [#5022 (comment)](https://github.com/kubernetes/enhancements/pull/5022#issuecomment-2627805687) before we continue on with this KEP.

https://github.com/kubernetes/kubernetes/pull/129922
As a short term fix, we can think about something like this

dchen1107 commented:
cc/ @SergeyKanzhelev can we treat this as part of node reliability enhancement project? cc/ @yujuhong for vis. too.

harshanarayana commented:
xref: https://github.com/kubernetes/kubernetes/issues/129975

More items related to the log rotate workflows.

yujuhong commented:
Adding @leonzz to help review the potential impact on the log collection pipeline.

Zeel-Patel commented:
> @Zeel-Patel is the KEP up-to-date with your current thinking?
> 
> Design details talk about imagefs/nodefs.

Hi @kannon92 , sorry, missed. Updated the KEP and added more details. Please re-review

leonzz commented on keps/sig-node/5032-log-eviction-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,318 @@
+3. It does cleanup to delete all non-rotated (rotated logs have time suffix in name) and .tmp files generated (and if not deleted) in last log rotation.
+4. Then it rotates un rotated files (it does not rotate the active file) and deletes oldest rotated files till containerLogMaxFiles-2 files are left. This is because the non rotated active file be rotated and new active file will be created. Which will add upto containerLogMaxFiles.
+5. Before doing this exact above step, in the new design, it will split the large active log file in size of containerLogMaxSize and name them <active-log-file-name>.part<i>.
+6. Let's say it created n parts, after this, it can do rotate of n-1 parts, keeping last nth part active and do delete. (Basically step 4)
```
> Adding @leonzz to help review the potential impact on the log collection pipeline.

> keeping last nth part active

If you truncate a 11M `container.log` in place to 1M and copy out the first 10M to `container.part1`,  it could potentially mess up the read position with some log collectors. For example fluentbit would re-read the 1M content after the truncation, resulting in duplicated logs, because it would always read from beginning when truncation is detected (see the discussion in https://github.com/fluent/fluent-bit/issues/375#issuecomment-363040116)

For log collectors to behave correctly, it's better to just create a new empty active file but this means the nth part of the last active file could be much smaller than containerLogMaxSize

Zeel-Patel replied:
Could we think about keeping the tailing logs to the last part itself till the size is about containerLogMaxSize/2.
Create a new part for tailing logs only if it more than containerLogMaxSize/2.

And create entirely new active file always.

Zeel-Patel replied:
@leonzz 
Can we think about keeping the tailing logs to the last part itself till the size is about containerLogMaxSize/2.
Create a new part for tailing logs only if it is more than containerLogMaxSize/2.

And create entirely new active file always.

leonzz replied:
> Create a new part for tailing logs only if it is more than containerLogMaxSize/2.

To me this sounds like a reasonable tradeoff to `always create new active files`.

So we could have files sizes anywhere in between 0.5 to 1.5 times containerLogMaxSize, which I assume can still resolve the problems you encounter with super large files?

sftim commented on keps/sig-node/5032-log-eviction-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,319 @@
+
+## Motivation
+
+- We manage kubernetes ecosystem at our organization. A lot of our kubelet hosts experienced Disk pressure as a certain set of pods was generating very high logs. The rate was around 3-4Gib logs generated in 15 minutes. We had kubelet configs containerLogMaxSize set to 200Mib and containerLogMaxFiles set to 6. But the .gz files (tar log files of pods) were of size around 500-600Gib. We observed t ...
```
This is very specific and doesn't represent Kubernetes as an organization. @Zeel-Patel you should reword this motivation to cover the _Kubernetes project_ motivation to make this change.

sftim commented on keps/sig-node/5032-log-eviction-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,319 @@
+
+### What would you like to be added?
+
+We expect that the log file size is always under set kubelet config limit (i.e., containerLogMaxSize) which can help such disk pressure issues in the future.
```
In this paragraph, who does “we“ refer to?

sftim commented on keps/sig-node/5032-log-eviction-on-disk-pressure/README.md:
````diff
@@ -0,0 +1,319 @@
+containerLogMaxFiles = 6
+```
+
+### Spec 1
+
+Continuously generating 10Mib with 0.1 sec sleep in between
````
```suggestion
### Example 1

This manifest defines a Job that continuously generates 10MiB of logs with 0.1 sec sleep in between (100MiB/s):
```

sftim commented on keps/sig-node/5032-log-eviction-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,319 @@
+
+### Risks and Mitigations
+
+Do not see any risk as of now.
```
Are you sure?

sftim commented on keps/sig-node/5032-log-eviction-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,319 @@
+
+## Design Details
+
+1. Implement a new function (splitAndRotateLatestLog) to be called by rotateLog function (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/logs/container_log_manager.go#L313-L346)
```
```suggestion
1. Implement a new function (`splitAndRotateLatestLog()`) to be called by the `rotateLog` function (https://github.com/kubernetes/kubernetes/blob/9e555875e79f60a238952e91e91f79f32c053f9c/pkg/kubelet/logs/container_log_manager.go#L313-L346)
2. ```

sftim commented on keps/sig-node/5032-log-eviction-on-disk-pressure/README.md:
```diff
@@ -0,0 +1,319 @@
+
+
+###### Are there any tests for feature enablement/disablement?
+Add UTs.
```
```suggestion

- Unit tests will be added for alpha

```

?

sftim reviewed (commented):
Here's some feedback. I'm not sure what quality level we expect for provisional KEPs, but I hope it's helpful even if it's optional to accept.

haircommander commented:
@Zeel-Patel were you planning on pushing forward on this in 1.34?
